### PR TITLE
Support init containers

### PR DESCRIPTION
**Pull request by fernandoacorreia**
Adds support for optionally specifying one or more init containers that will run before the main build container.

These init containers will have access to the shared build context volume, so that files created by them can be part of the final Docker image.

See `config/samples/init-container-sample.yaml` for an example. Relevant part:

```
  initContainers:
    - image: "busybox"
      command: ["/bin/sh"]
      args:
        - "-c"
        - "mkdir -p $TARGET_DIR && chmod 0777 $TARGET_DIR && echo $(date) $MESSAGE > $TARGET_DIR/init-container-output.txt"
      env:
        - name: MESSAGE
          value: "Hello from init container"
        - name: TARGET_DIR
          value: "/mnt/build/extracted"
```

Sample output:

```
❯ minikube ssh
Last login: Tue Mar  2 22:48:19 2021 from 192.168.49.1
docker@minikube:~$ docker pull localhost:32002/init-container-sample
Using default tag: latest
latest: Pulling from init-container-sample
ba3557a56b15: Already exists
54b0d889b94a: Already exists
ef397b6c435a: Already exists
7ba08b745588: Already exists
3eebe66e00d3: Already exists
059138af06d1: Extracting [>                                                  ]  65.54kB/4.464MB
059138af06d1: Pull complete
777428c54d84: Pull complete
4c1e475b2932: Pull complete
Digest: sha256:9bb5ee2ef208ba721a7bd5ea9777b9c395ce12c3eb7002349d24306867c324b2
Status: Downloaded newer image for localhost:32002/init-container-sample:latest
localhost:32002/init-container-sample:latest
docker@minikube:~$
docker@minikube:~$ docker run --rm -it --entrypoint="/bin/sh" localhost:32002/init-container-sample
/app # cat
Dockerfile                 app.py                     init-container-output.txt
/app # cat init-container-output.txt
Tue Mar 2 22:51:42 UTC 2021 Hello from init container
```

Notes: 

1. Forge uses the `extracted` directory as the build context. On a subsequent PR I think that we might want to refactor this to a more neutral name like `/mnt/build/context` or at least make it configurable via the CIB.

2. There's an extra volume mount being added to the init containers. Ideally that should be suppressed so that users with CIB access can't use Forge service account tokens.
```
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: build-init-container-sample-token-8kmwt
      readOnly: true
```

### CODE REVIEW

**Review by ddl-deanwampler — APPROVED** [review]
LGTM, but I'm not the expert ;)

File: e2e/run-test.sh [hunk 1]
```diff
@@ -53,28 +53,33 @@ function run_test {
...
     error  "diff failed"
-    ls -lahR "$test_dir"
+    echo "EXPECTED:"
+    ls -lahR "$expected_dir"
+    echo "ACTUAL:"
+    ls -lahR "$actual_dir"
     exit 1
   fi
 
-  info "Test succeeded"
```
Comment (sonnysideup) [inline e2e/run-test.sh:77]: Don't we want to keep this line albeit after the following one so we can easily distinguish when an e2e test passes? I think you and I would have a sense that `Files in ...` means the test passes but new devs probably would not, right?

File: e2e/run-test.sh [hunk 2]
```diff
@@ -174,10 +179,17 @@ run_test "Build should push to a private registry with TLS enabled" \
           e2e/builds/tls_with_basic_auth.yaml \
           test-tls-with-basic-auth \
           "$namespace"
+
 run_test "Build should pull base image from a private registry" \
           e2e/builds/private_base_image.yaml \
           test-private-base-image \
           "$namespace"
-verify_image
+verify_image variable-base-app /app $BASE_DIR/internal/archive/testdata/expected/simple-app
```
Comment (sonnysideup) [inline e2e/run-test.sh:187]: I would move `interna/archive/testdata/{simple-app,init-container}` out of that directory and  nested it under `./e2e`. The former indicates it's test data for the "archive" tests which it is not.

**Review by sonnysideup — APPROVED** [review]
@fernandoacorreia LGTM but please check out my comments. There's nothing wrong here but I did make some suggestions that I think are valid.

File: e2e/run-test.sh (hunk 1 above)
Comment (fernandoacorreia) [reply e2e/run-test.sh:77]: good idea

File: e2e/run-test.sh (hunk 2 above)
Comment (fernandoacorreia) [reply e2e/run-test.sh]: done

**Pull request merged**